File: fp/tests/test_selection.py
import sys
from pathlib import Path

# Add parent directory to path so we can import from src
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from ga_operators import tournament_select
from type_alias import Population
from fake_random_gen import fake_random_engine


class TestSelection:
    """Test suite for Tournament Selection operator"""
    
    @staticmethod
    def test_01():
        """Test 01: Tournament selection with size 3 from population of 5
        Tests: Selects the best chromosome from tournament (indices 0, 3, 4)
        """
        population: Population = [
            ([False] * 10, 10.0),
            ([False] * 10, 50.0),
            ([False] * 10, 30.0),
            ([True] * 10, 80.0),  # Best: index 3 with fitness 80.0
            ([False] * 10, 20.0),
        ]
        
        # Tournament selects from indices 0, 3, 4 -> best is index 3 with fitness 80.0
        rand_eng = fake_random_engine([0, 3, 4])
        selected = tournament_select(population, 3, rand_eng=rand_eng)
        
        assert selected is not None, "Should select a chromosome"
        assert selected == [True] * 10, "Should return the best chromosome (all True)"
    
    @staticmethod
    def test_02():
        """Test 02: Tournament selection - verify correct chromosome selection with different indices
        Tests: Selects best chromosome from different tournament participants
        """
        population: Population = [
            ([False] * 10, 25.0),
            ([False] * 10, 15.0),
            ([True] * 10, 90.0),  # Best: index 2 with fitness 90.0
            ([False] * 10, 40.0),
            ([False] * 10, 60.0),
            ([False] * 10, 35.0),
        ]
        
        # Tournament selects from indices 1, 2, 4, 5 -> best is index 2 with fitness 90.0
        rand_eng = fake_random_engine([1, 2, 4, 5])
        selected = tournament_select(population, 4, rand_eng=rand_eng)
        
        assert selected is not None, "Should select a chromosome"
        assert selected == [True] * 10, "Should return the best chromosome (all True)"
    
    @staticmethod
    def test_03():
        """Test 03: Tournament selection - single participant tournament
        Tests: Correctly returns that single participant
        """
        population: Population = [
            ([False] * 10, 10.0),
            ([False] * 10, 85.0),
            ([True] * 10, 30.0),  # Selected: index 2
            ([False] * 10, 50.0),
            ([False] * 10, 20.0),
        ]
        
        # Tournament size 1 selects index 2 -> returns that chromosome
        rand_eng = fake_random_engine([2])
        selected = tournament_select(population, 1, rand_eng=rand_eng)
        
        assert selected is not None, "Should select a chromosome"
        assert selected == [True] * 10, "Should return the chromosome at index 2"
    
    @staticmethod
    def test_04():
        """Test 04: Tournament selection - edge cases
        Tests: Tournament size equals population size, same fitness values, large populations
        """
        population: Population = [
            ([False] * 10, 50.0),
            ([False] * 10, 50.0),
            ([True] * 10, 100.0),  # Best: index 2 with fitness 100.0
            ([False] * 10, 50.0),
        ]
        
        # Edge case 1: Tournament size equals population size
        # Selects from indices 0, 1, 2, 3 -> best is index 2 with fitness 100.0
        rand_eng = fake_random_engine([0, 1, 2, 3])
        selected = tournament_select(population, 4, rand_eng=rand_eng)
        assert selected is not None, "Should select a chromosome"
        assert selected == [True] * 10, "Should return the best chromosome (all True)"
        
        # Edge case 2: Tournament size 2 with duplicate fitnesses
        # Selects from indices 1, 3 -> both have fitness 50.0
        rand_eng = fake_random_engine([1, 3])
        selected = tournament_select(population, 2, rand_eng=rand_eng)
        assert selected is not None, "Should select a chromosome"
        assert selected == [False] * 10, "Should return one of the selected chromosomes"
        
        # Edge case 3: Large tournament size from large population
        large_pop: Population = [
            ([False] * 10, float(i * 2.0)) if i != 45 else ([True] * 10, 90.0)
            for i in range(50)
        ]
        # Selects from indices 5, 15, 25, 35, 45, 8, 12, 22, 30, 40
        # Best is index 45 with fitness 90.0
        rand_eng = fake_random_engine([5, 15, 25, 35, 45, 8, 12, 22, 30, 40])
        selected = tournament_select(large_pop, 10, rand_eng=rand_eng)
        assert selected is not None, "Should select a chromosome"
        assert selected == [True] * 10, "Should return the best chromosome (all True)"
    
    @staticmethod
    def test_05():
        """Test 05: Tournament selection - all same fitness values
        Tests: Selection with identical fitness values
        """
        population: Population = [
            ([False] * 10, 75.0),
            ([True] * 10, 75.0),  # This one is all True
            ([False] * 10, 75.0),
        ]
        
        # All chromosomes have same fitness, selection from indices 0, 2
        rand_eng = fake_random_engine([0, 2])
        selected = tournament_select(population, 2, rand_eng=rand_eng)
        
        assert selected is not None, "Should select a chromosome"
        assert selected in [[False] * 10, [True] * 10], "Should return one of the tournament participants"
        assert population[0][0] == [False] * 10, "FP: Original population not modified by selection"


def run_all_tests():
    """Run all test methods"""
    print("Running Selection tests...")
    TestSelection.test_01()
    print("  ✓ test_01: Tournament size 3 from 5")
    TestSelection.test_02()
    print("  ✓ test_02: Different indices")
    TestSelection.test_03()
    print("  ✓ test_03: Single participant")
    TestSelection.test_04()
    print("  ✓ test_04: Edge cases")
    TestSelection.test_05()
    print("  ✓ test_05: Same fitness values")
    
    print("\n✅ All selection tests passed!")


if __name__ == "__main__":
    run_all_tests()

File: oop/tests/strategies/TestSelection.java
package strategies;

import models.*;
import utils.RandomUtil;
import utils.FakeRandomUtil;

/**
 * javac -d bin src/models/*.java src/utils/*.java src/strategies/*.java tests/utils/FakeRandomUtil.java tests/strategies/TestSelection.java
 * java -ea -cp bin strategies.TestSelection
 */
public class TestSelection {
    /**
     * Test 01: Tournament selection with size 3 from population of 5
     * Tests: Selects the best chromosome from tournament (indices 0, 3, 4)
     */
    public static void test_01() {
        Chromosome[] chromosomes = new Chromosome[5];
        chromosomes[0] = new Chromosome(new boolean[10], 10.0);
        chromosomes[1] = new Chromosome(new boolean[10], 50.0);
        chromosomes[2] = new Chromosome(new boolean[10], 30.0);
        chromosomes[3] = new Chromosome(new boolean[10], 80.0);
        chromosomes[4] = new Chromosome(new boolean[10], 20.0);
        
        Population pop = new Population(chromosomes, 5);
        TournamentSelection tournament = new TournamentSelection(3);
        
        // FakeRandomUtil will return 0, 3, 4 in sequence
        FakeRandomUtil fakeRandom = new FakeRandomUtil(new int[]{0, 3, 4}, new boolean[]{});
        
        Chromosome selected = tournament.select(pop, fakeRandom);
        
        assert selected.getFitness() == 80.0 : "Should select chromosome at index 3 with fitness 80.0 (max of 10.0, 80.0, 20.0)";
        assert selected == chromosomes[3] : "Should select the exact chromosome object from index 3";
    }
    
    /**
     * Test 02: Tournament selection - verify correct chromosome selection with different indices
     * Tests: Selects best chromosome from different tournament participants
     */
    public static void test_02() {
        Chromosome[] chromosomes = new Chromosome[6];
        chromosomes[0] = new Chromosome(new boolean[10], 25.0);
        chromosomes[1] = new Chromosome(new boolean[10], 15.0);
        chromosomes[2] = new Chromosome(new boolean[10], 90.0);
        chromosomes[3] = new Chromosome(new boolean[10], 40.0);
        chromosomes[4] = new Chromosome(new boolean[10], 60.0);
        chromosomes[5] = new Chromosome(new boolean[10], 35.0);
        
        Population pop = new Population(chromosomes, 6);
        TournamentSelection tournament = new TournamentSelection(4);
        
        // FakeRandomUtil will return 1, 2, 4, 5 in sequence
        FakeRandomUtil fakeRandom = new FakeRandomUtil(new int[]{1, 2, 4, 5}, new boolean[]{});
        
        Chromosome selected = tournament.select(pop, fakeRandom);
        
        assert selected.getFitness() == 90.0 : "Should select chromosome at index 2 with fitness 90.0 (max of 15.0, 90.0, 60.0, 35.0)";
        assert selected == chromosomes[2] : "Should select the exact chromosome object from index 2";
    }
    
    /**
     * Test 03: Tournament selection - single participant tournament
     * Tests: Correctly returns that single participant
     */
    public static void test_03() {
        Chromosome[] chromosomes = new Chromosome[5];
        chromosomes[0] = new Chromosome(new boolean[10], 10.0);
        chromosomes[1] = new Chromosome(new boolean[10], 85.0);
        chromosomes[2] = new Chromosome(new boolean[10], 30.0);
        chromosomes[3] = new Chromosome(new boolean[10], 50.0);
        chromosomes[4] = new Chromosome(new boolean[10], 20.0);
        
        Population pop = new Population(chromosomes, 5);
        TournamentSelection tournament = new TournamentSelection(1);
        
        // FakeRandomUtil will return 2
        FakeRandomUtil fakeRandom = new FakeRandomUtil(new int[]{2}, new boolean[]{});
        
        Chromosome selected = tournament.select(pop, fakeRandom);
        
        assert selected.getFitness() == 30.0 : "Should select chromosome at index 2 with fitness 30.0";
        assert selected == chromosomes[2] : "Should select the exact chromosome object from index 2";
    }
    
    /**
     * Test 04: Tournament selection - edge cases
     * Tests: Tournament size equals population size, same fitness values, all chromosomes valid
     */
    public static void test_04() {
        Chromosome[] chromosomes = new Chromosome[4];
        chromosomes[0] = new Chromosome(new boolean[10], 50.0);
        chromosomes[1] = new Chromosome(new boolean[10], 50.0);
        chromosomes[2] = new Chromosome(new boolean[10], 100.0);
        chromosomes[3] = new Chromosome(new boolean[10], 50.0);
        
        Population pop = new Population(chromosomes, 4);
        
        // Edge case 1: Tournament size equals population size
        TournamentSelection tournament4 = new TournamentSelection(4);
        FakeRandomUtil fakeRandom = new FakeRandomUtil(new int[]{0, 1, 2, 3}, new boolean[]{});
        Chromosome selected = tournament4.select(pop, fakeRandom);
        assert selected.getFitness() == 100.0 : "Should select best from entire population (index 2 with 100.0)";
        assert selected == chromosomes[2] : "Should select the exact chromosome object from index 2";
        
        // Edge case 2: Tournament size 2 with duplicate fitnesses
        TournamentSelection tournament2 = new TournamentSelection(2);
        FakeRandomUtil fakeRandom2 = new FakeRandomUtil(new int[]{1, 3}, new boolean[]{});
        Chromosome selected2 = tournament2.select(pop, fakeRandom2);
        assert selected2.getFitness() == 50.0 : "Should select one of the chromosomes with 50.0 fitness";
        assert (selected2 == chromosomes[1] || selected2 == chromosomes[3]) : "Should be one of the selected indices";
        
        // Edge case 3: Large tournament size from large population
        Chromosome[] largeChromosomes = new Chromosome[50];
        for (int i = 0; i < 50; i++) {
            largeChromosomes[i] = new Chromosome(new boolean[10], (double) i * 2.0);
        }
        Population largePop = new Population(largeChromosomes, 50);
        TournamentSelection tournament10 = new TournamentSelection(10);
        int[] indices = {5, 15, 25, 35, 45, 8, 12, 22, 30, 40};
        FakeRandomUtil fakeRandom3 = new FakeRandomUtil(indices, new boolean[]{});
        Chromosome selected3 = tournament10.select(largePop, fakeRandom3);
        assert selected3.getFitness() == 45 * 2.0 : "Should select index 45 with fitness 90.0 (max of provided indices)";
        assert selected3 == largeChromosomes[45] : "Should select the exact chromosome object from index 45";
    }
    
    /**
     * Test 05: Tournament selection - boundary and invalid input cases
     * Tests: Tournament size > population size throws exception, all same fitness values
     */
    public static void test_05() {
        Chromosome[] chromosomes = new Chromosome[3];
        chromosomes[0] = new Chromosome(new boolean[10], 75.0);
        chromosomes[1] = new Chromosome(new boolean[10], 75.0);
        chromosomes[2] = new Chromosome(new boolean[10], 75.0);
        
        Population pop = new Population(chromosomes, 3);
        
        // Edge case 1: Tournament size > population size should throw
        TournamentSelection tournament5 = new TournamentSelection(5);
        FakeRandomUtil fakeRandom = new FakeRandomUtil(new int[]{0, 1, 2, 0, 1}, new boolean[]{});
        boolean exceptionThrown = false;
        try {
            tournament5.select(pop, fakeRandom);
        } catch (IllegalArgumentException e) {
            exceptionThrown = true;
            assert e.getMessage().contains("tournamentSize") : "Exception message should mention tournamentSize";
        }
        assert exceptionThrown : "Should throw IllegalArgumentException when tournament size > population size";
        
        // Edge case 2: All chromosomes have same fitness - any selection is valid
        TournamentSelection tournament2 = new TournamentSelection(2);
        FakeRandomUtil fakeRandom2 = new FakeRandomUtil(new int[]{0, 2}, new boolean[]{});
        Chromosome selected = tournament2.select(pop, fakeRandom2);
        assert selected.getFitness() == 75.0 : "Should select a chromosome with fitness 75.0";
        assert (selected == chromosomes[0] || selected == chromosomes[2]) : "Should be one of the tournament participants";
        
        // Edge case 3: Very small population (size 2)
        Chromosome[] smallChromosomes = new Chromosome[2];
        smallChromosomes[0] = new Chromosome(new boolean[10], 10.0);
        smallChromosomes[1] = new Chromosome(new boolean[10], 99.0);
        Population smallPop = new Population(smallChromosomes, 2);
        TournamentSelection tournament2_small = new TournamentSelection(2);
        FakeRandomUtil fakeRandom3 = new FakeRandomUtil(new int[]{0, 1}, new boolean[]{});
        Chromosome selected2 = tournament2_small.select(smallPop, fakeRandom3);
        assert selected2.getFitness() == 99.0 : "Should select best from 2-chromosome population";
        assert selected2 == smallChromosomes[1] : "Should select chromosome at index 1";
    }
    
    public static void main(String[] args) {
        test_01();
        test_02();
        test_03();
        test_04();
        test_05();
        System.out.println("All tournament selection tests passed!");
    }
}